File: src/net/dinkla/raytracer/math/Basis.java
//package net.dinkla.raytracer.math;
//
///**
// * Created by IntelliJ IDEA.
// * User: jorndinkla
// * Date: 07.06.2010
// * Time: 19:44:08
// * To change this template use File | Settings | File Templates.
// */
//public class Basis {
//
//    public Vector3DF u, v, w;
//
//    public Basis(final Point3DF eye, final Point3DF lookAt, final Vector3DF up) {
//        w = eye.minus(lookAt).normalize();
//        u = up.cross(w).normalize();
//        v = w.cross(u);
//    }
//
//    public Vector3DF pm(final float x, final float y, final float z) {
//        return u.mult(x).plus(v.mult(y)).minus(w.mult(z));
//    }
//
//    public Vector3DF pp(final float x, final float y, final float z) {
//        return u.mult(x).plus(v.mult(y)).plus(w.mult(z));
//    }
//
//}

File: src/net/dinkla/raytracer/math/Basis.scala
package net.dinkla.raytracer.math

/**
 * Created by Dinkla on 10.07.2015.
 */
class Basis(eye: Point3DF, lookAt: Point3DF, up: Vector3DF) {

  lazy val u: Vector3D[Float] = up.cross(w).normalize
  lazy val v: Vector3D[Float] = w.cross(u)
  lazy val w: Vector3D[Float] = eye.minus(lookAt).normalize

  def pm(x: Float, y: Float, z: Float): Vector3D[Float]
    = u * x + v * y - w * z

  def pp(x: Float, y: Float, z: Float): Vector3D[Float]
    = u * x + v * y + w * z

}
